gematria/datasets/exegesis_benchmark_lib.h

```diff
@@ -0,0 +1,61 @@
+// Copyright 2024 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <cstddef>
+#include <memory>
+#include <string_view>
+
+#include "gematria/datasets/find_accessed_addrs.h"
+#include "llvm/MC/MCAsmInfo.h"
+#include "llvm/MC/MCDisassembler/MCDisassembler.h"
+#include "llvm/MC/MCInst.h"
+#include "llvm/MC/MCInstPrinter.h"
+#include "llvm/Support/Error.h"
+#include "llvm/Support/JSON.h"
+#include "llvm/tools/llvm-exegesis/lib/BenchmarkCode.h"
+#include "llvm/tools/llvm-exegesis/lib/BenchmarkRunner.h"
+#include "llvm/tools/llvm-exegesis/lib/LlvmState.h"
+
+namespace gematria {
+
+// ExegesisBenchmark is used for generating benchmark configurations for
+// benchmarks and executing those benchmarks to get real-world data that
+// can then be used for training Gematria models.
+class ExegesisBenchmark {
+ private:
+  explicit ExegesisBenchmark(llvm::exegesis::LLVMState &&State);
```

Comment:
Do we need to take the state by rvalue reference? This might complicate memory management in the Python binding (I'm not sure how well a move translates, and it might create dangling pointers or accessible moved objects).

I think we should use raw pointers, and something like https://pybind11.readthedocs.io/en/stable/advanced/functions.html#keep-alive to make sure that the Python object doesn't get garbage collected.

Reply:
This constructor is private and only called from within the C++, particularly the `create` function, so I don't think we should have any problems with memory management here.